feat(curriculum): add project board lab (#56767)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/25-front-end-development/lab-project-idea-board/67051431a73f1ca25d9a6f25.md

---
id: 67051431a73f1ca25d9a6f25
title: Build a Project Idea Board
challengeType: 14
dashedName: build-a-project-idea-board
---

# --description--

**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. 

**User Stories:**

1. You should define an object constant named `projectStatus` with the following project statuses and descriptions:

   - `PENDING` with a description of `Pending Execution`.
   - `SUCCESS` with a description of `Executed Successfully`.
   - `FAILURE` with a description of `Execution Failed`.

> 2. You should define a class named `ProjectIdea` with a `constructor` that takes a `title` and a `description` as parameters and initialize the `title` and `description` properties based on the parameters passed. The class should also have a property named `status` and be set to the description of the `PENDING` status.

Reviewer: ```suggestion
2. You should define a class named `ProjectIdea` with a `constructor` that takes `title` and `description` as parameters. Initialize the `title` and `description` properties with the provided parameters. The class should also have a property named `status` that is set to the value `PENDING` by default.
```

Reviewer 2: ```suggestion
2. You should define a class named `ProjectIdea` with a `constructor` that takes `title` and `description` as parameters. Initialize the `title` and `description` properties with the provided parameters. The class should also have a property named `status` that is set to the value `projectStatus.PENDING` by default.
```

> 3. You should define a method named `updateProjectStatus` inside the `ProjectIdea` class that takes accepts `newStatus` as parameter and can update the project idea status.

Reviewer: ```suggestion
3. You should define a method named `updateProjectStatus` inside the `ProjectIdea` class. This method should accept a `newStatus` parameter and update the `status` property to the given value.
```

4. You should define a `ProjectIdeaBoard` class with a `constructor` that accepts a `title` and initializes an empty array named `ideas` to hold instances of the `ProjectIdea` class.
5. You should define a method named `pin` inside the `ProjectIdeaBoard` class that accepts an instance of the `ProjectIdea` class and pushes the given instance to the `ideas` array.

> 6. You should define a method named `unpin` inside the `ProjectIdeaBoard` class that accepts an instance of the `ProjectIdea` class and removes the given instance from the `ideas` array.

Reviewer: ```suggestion
6. You should define a method named `unpin` inside the `ProjectIdeaBoard` class. This method should accept an instance of the `ProjectIdea` class and removes it from the `ideas` array.
```

7. You should define a method named `count` that returns the number of project ideas in the given `ProjectIdeaBoard` array.

> 8. You should define a method named `toString` that returns the name of the projects ideas, their description and status in the format:

Author: done!

```js
[ProjectIdeaBoard title] has [ProjectIdeaBoard count] idea(s)\n
[ProjectIdea Title] ([ProjectIdea Status Description]) - [ProjectIdea Description]

> [ProjectIdea Title] ([ProjectIdea Status Description]) - [ProjectIdea Description]

Author: I remember we agreed on using square brackets for placeholders, like step#2 [here](https://github.com/freeCodeCamp/freeCodeCamp/blob/main/curriculum/challenges/english/25-front-end-development/lab-lunch-picker-program/66db529d37ad966480ebb633.md).

> [ProjectIdea Title] ([ProjectIdea Status Description]) - [ProjectIdea Description]

Reviewer 2: In any case we are using both types of brackets https://github.com/freeCodeCamp/freeCodeCamp/blob/4cc7c54979f5c7715603648f4f497223a87e168f/curriculum/challenges/english/25-front-end-development/lab-date-conversion/66f686b8ebdb982fa8e14330.md?plain=1#L43

I prefer angular brackets but it would be okay as long as we are consistent. I would like to hear other thoughts on this.

> [ProjectIdea Title] ([ProjectIdea Status Description]) - [ProjectIdea Description]

Reviewer 3: I think angular brackets are not a symbol for data types in JS so it is less confusing

> [ProjectIdea Title] ([ProjectIdea Status Description]) - [ProjectIdea Description]

Author: I have used angular brackets now.

.
.
.
```

# --hints--

You should define an object constant named `projectStatus` with the correct project statuses and descriptions:

```js
assert.isObject(projectStatus);
assert.sameMembers(Object.keys(projectStatus), ["PENDING", "SUCCESS", "FAILURE"]);
assert.deepEqual(projectStatus.PENDING, { description: "Pending Execution" });
assert.deepEqual(projectStatus.SUCCESS, { description: "Executed Successfully" });
assert.deepEqual(projectStatus.FAILURE, { description: "Execution Failed" });
```

You should have a `ProjectIdea` class.

```js
assert.isFunction(ProjectIdea);
assert.isObject(new ProjectIdea());
```

> Your `ProjectIdea` class should initialize the `title` property and `description` property based on the parameters passed. It should also set the status to the description of the `PENDING` status.

Reviewer 2: ```suggestion
Your `ProjectIdea` class should initialize the `title` property and `description` property based on the parameters passed. It should also set the status to `projectStatus.PENDING`.
```

breakfastChef.updateProjectStatus(projectStatus.FAILURE);

assert.deepEqual(breakfastChef, {
  title: "Breakfast Chef Robot",
  description:
    "A robot that can follow a given list of instructions and prepare breakfast for the user and let them know through their phone.",
  status: { description: "Execution Failed" },
});
```

Calling `updateProjectStatus(projectStatus.SUCCESS)` on `new ProjectIdea("Online Used Video Games Store", "An online platform where users can buy second hand physical copies of video games from other users.")` should update the status to `{ description: 'Executed Successfully' }`.

```js
let usedVideoGamesStore = new ProjectIdea(
  "Online Used Video Games Store",
  "An online platform where users can buy second hand physical copies of video games from other users."
);

// Update the project status set to SUCCESS
usedVideoGamesStore.updateProjectStatus(projectStatus.SUCCESS);

assert.deepEqual(usedVideoGamesStore, {
  title: "Online Used Video Games Store",
  description:
    "An online platform where users can buy second hand physical copies of video games from other users.",
  status: { description: "Executed Successfully" },
});
```

> When you pin `new ProjectIdea("Smart Home System", "An integrated system to control lighting, temperature, and security devices remotely.")` to `new ProjectIdeaBoard("Tech Projects Board")`, you should get `Tech Projects Board has 1 idea(s)\nSmart Home System (Pending Execution) - An integrated system to control lighting, temperature, and security devices remotely.`

Author: I have reworded this a bit.

> When you pin `new ProjectIdea("Smart Home System", "An integrated system to control lighting, temperature, and security devices remotely.")` to `new ProjectIdeaBoard("Tech Projects Board")`, your project board should show `Tech Projects Board has 1 idea(s)\nSmart Home System (Pending Execution) - An integrated system to control lighting, temperature, and security devices remotely.`

Reviewer 2: This is still confusing. When you use `pin`, you push a `ProjectIdea` to `ProjectIdeaBoard.ideas`. That's it.

It shouldn't be correlated to the result of calling `formatToString` after pinning an object.

`formatToString` should be tested in a separate test, setting the `ideas` array to a known content and checking the output.

Here I would simply test that calling `pin` adds the object to the `ideas` array. The same holds for `unpin`.

Author: Now, I have updated the hints and tested the inputs againtst the `ideas` array.

```
You should be able to pin a `ProjectIdea` object to your `ProjectIdeaBoard` using the `pin` method.
```
```
You should be able to unpin a `ProjectIdea` object to your `ProjectIdeaBoard` using the `unpin` method.
```
